src/ln/keysinterface.rs

```diff
@@ -0,0 +1,186 @@
+	/// BOTL 3 derivation scheme.
+	fn handle_spendable_output(&self, event: events::Event);
+}
+
+/// Hacky custom output script descriptors to ease spending of onchain outputs by user wallet
+/// Maybe should be changed by real ones when merged into rust-bitcoin.
+/// StaticOutputs commit to a static pubkey, i.e one derived once for node operation lifetime.
+/// DynamicOutputs commit to a dynamic local_delayedpubkey, i.e one which change for each per_commitment_point
+pub enum CustomOutputScriptDescriptor {
+	/// Outpoint commits to a P2PWKH, should be spend by the following witness :
+	/// <signature> <pubkey> OP_CHECKSIG
+	StaticOutput {
+		/// Outpoint spendable by user wallet
+		outpoint: OutPoint,
+	},
+	/// Outpoint commits to a P2WSH, should be spend by the following witness :
+	/// <local_delayedsig> 0 <witnessScript>
+	/// With input nSequence set to_self_delay.
+	DynamicOutput {
+		/// Outpoint spendable by user wallet
+		outpoint: OutPoint,
+		/// local_delayedkey = delayed_payment_basepoint_secret + SHA256(per_commitment_point || delayed_payment_basepoint
+		local_delayedkey: SecretKey,
+		/// witness script encumbering output
+		witness_script: Script,
+		/// nSequence input must commit to self_delay to satisfy script's OP_CSV
+		to_self_delay: u16,
+	}
+}
+
+impl CustomOutputScriptDescriptor {
+	/// Build a StaticOuput descriptor
+	pub fn static_key(outpoint: OutPoint) -> Self {
+		CustomOutputScriptDescriptor::StaticOutput {
+			outpoint,
+		}
+	}
+
+	/// Build a DynamicOuput descriptor
+	pub fn dynamic_key(outpoint: OutPoint, local_delayedkey: SecretKey, witness_script: Script, to_self_delay: u16) -> Self {
+		CustomOutputScriptDescriptor::DynamicOutput {
+			outpoint,
+			local_delayedkey,
+			witness_script,
+			to_self_delay,
+		}
+	}
+}
+
+/// A trait to describe an object which should get secrets from user wallet and apply derivation
+/// to provide keys materials downstream
+pub trait KeysInterface: Send + Sync {
+	/// User wallet has to provide this KeysInterface with a master seed used to derive 
+	/// node_id as /0' and destination_script as /N'
+	fn provide_master_seed(&mut self, seed : &[u8; 32]);
+	/// Get node secret key to derive node_id
+	fn get_node_secret(&self) -> Option<SecretKey>;
+	/// Get destination redeemScript to encumber static protocol exit points. For now
+	/// redeemScript is a pay-2-public-key-hash.
+	fn get_destination_script(&self) -> Option<Script>;
+	/// Get shutdown_pubkey to use as PublicKey at channel closure
+	fn get_shutdown_pubkey(&self) -> Option<PublicKey>;
+	/// Get a new set of ChannelKeys from per-channel random key /3/N'
+	/// For Channel N, keys correspond to ChannelKeys::new_from_seed(/3/N')
+	fn get_channel_keys(&mut self) -> Option<ChannelKeys>;
+}
+
+/// Utility for storing/deriving lightning keys materials
+pub struct KeysManager {
+	network: Network,
+	secp_ctx: Secp256k1<secp256k1::All>,
+	master_key: Option<SecretKey>,
```

Comment:
Seems like it might be nicer to not make any of these Option<>al and make the interface only send data in one direction.

Reply:
Done also, there is the case of destination_script, maybe the new() should return KeysManager and destination priv key to let user wallet be in autonomy on this ?